ports/stm/common-hal/audiobusio/PDMIn.c

```diff
@@ -0,0 +1,199 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2019 Jeff Epler for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdint.h>
+#include "common-hal/audiobusio/PDMIn.h"
+#include "shared-bindings/audiobusio/PDMIn.h"
+#include "shared-bindings/microcontroller/Pin.h"
+#include "py/runtime.h"
+#include "supervisor/memory.h"
+#include "MEMS_Audio_ll_stm32l4.h"
+
+
+MemsAudio memsAudio;
+MemsAudio_STM32L4SAIPDM memsAudioImpl;
+pdm_sample_t pdmBuffer[MEMS_AUDIO_PDM_BUFFER_LENGTH];
+audiobusio_pdmin_obj_t *instance;
+
+static bool pdm_data_available(MemsAudio_STM32L4SAIPDM *impl, uint8_t *pdmBuffer, size_t pdmBufferLength);
+
+// Caller validates that pins are free.
+void common_hal_audiobusio_pdmin_construct(audiobusio_pdmin_obj_t *self,
+    const mcu_pin_obj_t *clock_pin,
+    const mcu_pin_obj_t *data_pin,
+    uint32_t sample_rate,
+    uint8_t bit_depth,
+    bool mono,
+    uint8_t oversample) {
+    self->sample_rate = sample_rate;
+    self->mono = mono;
+    self->oversample = oversample;
+    self->recording_complete = true;
+
+
+    if (!mono) {
+        mp_raise_ValueError(translate("only mono is supported"));
+    }
+    if (sample_rate != 16000) {
+        mp_raise_ValueError(translate("only sample_rate=16000 is supported"));
+    }
+    if (bit_depth != 16) {
+        mp_raise_ValueError(translate("only bit_depth=16 is supported"));
+    }
+    if (oversample != 64) {
+        mp_raise_ValueError(translate("only oversample=64 is supported"));
```

Comment:
Use the `py/argcheck.c` `mp_arg_validate...()` routines here instead, if possible. This allows reuse of existing error message, saving space, and saving work for the translators. Look at some other uses of `mp_arg_validate*` in `ports/*/common-hal/`.

If there is no existing routine you can use, at least try to reuse an existing error message. Use the `%q` (for QSTR) designator in the error message, so you can pass in an `MP_QSTR_...` arg name.

Reply:
I can look into the changes you suggested, thanks for the pointer.  I'd much prefer to use an established pattern than something ad hoc.

Just for some background, most of the error messages were present in the nrf port, so the translations appear to be reused - the translation file now has 2 file references for the same translation text that already existed. This should mean reduced work for the translators. I can look into quick wins to reduce this further.

Reply:
If that's the case, we can leave it alone for now, and then shrink it later. I was wondering if the STM32L4 had limited flash, as well.

Reply:
So don't bother changing these now, unless you need the space.

Reply:
That's good to hear. Thanks!

Reply:
Just to follow up on your question about flash space - the L4+ used in the Swan has 2MB flash, partitioned as

* 64kb - UF2 bootloader
* 960kb - runtime (presently 520kB used for a release build.)
* 1024kb - python scripts